### ISSUE TITLE
Can't get plugin rules added by `cli.addPlugin`

### ISSUE THREAD

**Issue by fisker** [eslint/eslint#12425]
**Tell us about your environment**

* **ESLint Version:** 6.5.1
* **Node Version:** 10.16.3
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**

default

```js
const {CLIEngine} = require('eslint')
const unicorn = require('eslint-plugin-unicorn')
const cli = new CLIEngine()
cli.addPlugin('eslint-plugin-unicorn', require('eslint-plugin-unicorn'))
console.log(cli.getRules().has('unicorn/no-process-exit'))
```

I can't get the plugin rules, Is this by design or a bug?

----

quick reproduction

```bash
md eslint-issues-12425 && cd eslint-issues-12425 && yarn init -y && yarn add eslint eslint-plugin-unicorn

# print true
node -p "new (require('eslint').CLIEngine)({plugins: ['unicorn']}).getRules().has('unicorn/no-process-exit')"

# print false, but should be true
node -p "const cli = new (require('eslint').CLIEngine);cli.addPlugin('eslint-plugin-unicorn', require('eslint-plugin-unicorn'));cli.getRules().has('unicorn/no-process-exit')"
```

**Comment by fisker** [comment]
Maybe related to https://github.com/eslint/eslint/issues/11871

**Comment by mysticatea** [comment]
Thank you for your report.

That's expected because `addPlugin()` actually just adds the implementation of a plugin. You still need `{plugins: ['unicorn']}` if you don't have `.eslintrc.*` files. We cannot use the rules of the plugins that the configuration doesn't have.

On the other hand, it looks like a bug because

```js
const { CLIEngine } = require('eslint')
const engine = new CLIEngine({ plugins: ["foo"] })

engine.addPlugin('foo', require('eslint-plugin-unicorn'))
console.log(engine.getRules().has('foo/no-process-exit'))
```

was false. I expected true in this case.

**Comment by mysticatea** [comment]
As a side note, I'm proposing to remove `addPlugin()` method in favor of a constructor option in RFC40.

**Comment by fisker** [comment]
The reason I check this, is because I want run plugin against plugin source, so I add cli to do this https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/test/lint/lint.js , I want make sure the plugin is loaded, but get a false result.

According to the comments above, I'm not doing it right.

I should point `eslint-plugin-unicorn` to local file and use `.eslintrc.*`, because we can't pass plugin path to `engine`, right?

**Comment by mysticatea** [comment]
In that case, you can use `getRules()` after the call of `executeOnFiles()` method for now.

The `getRules()` method returns the core rules and the rules of the plugins that the last call of `executeOnFiles()` or `executeOnText()` used.

Because, internally, the configuration object stores the loaded plugins that the config used. ESLint doesn't load plugins until configurations are determined and configurations are not determined until target files are determined. `CLIEngine` stores the configuration objects that the last call of `executeOnFiles()` or `executeOnText()`, then the `getRules()` method retrieves plugin rules from the configuration objects.

#11871 has updated `CLIEngine` to initialize "the configuration objects that the last call of `executeOnFiles()` or `executeOnText()` used" with only CLI options.

The `addPlugin()` looks to not update "the configuration objects that the last call of `executeOnFiles()` or `executeOnText()` used." (I think this is a bug because the method changes where ESLint loads plugins from.)
However, an `executeOnFiles()` call updates the configuration objects.

**Comment by fisker** [comment]
I remember I tried call this after `executeOnFiles`, but also got false. anyway will try later, not on laptop now

**Comment by fisker** [comment]
Tested

```text

$ node -p "const cli = new (require('eslint').CLIEngine);cli.addPlugin('eslint-plugin-unicorn', require('eslint-plugin-unicorn'));cli.getRules().has('unicorn/no-process-exit')"
false

$ node -p "const cli = new (require('eslint').CLIEngine)({useEslintrc: false});cli.addPlugin('eslint-plugin-unicorn', require('eslint-plugin-unicorn'));cli.executeOnText('');cli.getRules().has('unicorn/no-process-exit')"
false

**Comment by mysticatea** [comment]
> That's expected because addPlugin() actually just adds the implementation of a plugin. You still need {plugins: ['unicorn']} if you don't have .eslintrc.* files. We cannot use the rules of the plugins that the configuration doesn't have.
> https://github.com/eslint/eslint/issues/12425#issuecomment-542060798

**Comment by fisker** [comment]
Sorry, I miss understood it, my purpose is avoid use `{plugins: ['unicorn']}`, because this requires a `package.json` file has `eslint-plugin-*` as dependency.

Thanks a lot for explaining
